### PR TITLE
Implement Tokens in Swift and Kotlin

### PR DESCRIPTION
**Pull request by w11wo**
Implemented tokens in Swift and Kotlin API. Tested on iOS and Android.

### CODE REVIEW

File: sherpa-onnx/c-api/c-api.h [hunk 1]
```diff
@@ -101,8 +101,33 @@ SHERPA_ONNX_API typedef struct SherpaOnnxOnlineRecognizerConfig {
 } SherpaOnnxOnlineRecognizerConfig;
 
 SHERPA_ONNX_API typedef struct SherpaOnnxOnlineRecognizerResult {
+  // Recognized text
   const char *text;
-  // TODO(fangjun): Add more fields
+
+  // Pointer to continuous memory which holds string based tokens
+  // which are seperated by \0
+  const char *tokens;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h:109]: I suggest that we define
```
const char *_tokens;
const char *const*tokens;
```

where `_tokens` is your current `tokens`

And the new `char **tokens` is a pointer array.
`tokens[0]` contains the address of the first `token` in
`_tokens`. `tokens[1]` contains the address of the second
`token` in `_tokens`.

In this way, it simplifies users' life as they only need to iterate
`char **tokens`.

```
for (int32 i = 0; i != count; ++i) {
  const char*t = tokens[i];
  // process this token
}
```

Comment (ductranminh) [reply sherpa-onnx/c-api/c-api.h:109]: @csukuangfj  I added `const char *const *tokensArr;` for easy to implement, and I also tested

File: sherpa-onnx/c-api/c-api.cc [hunk 2]
```diff
@@ -154,18 +154,23 @@ SherpaOnnxOnlineRecognizerResult *GetOnlineStreamResult(
...
       pos += result.tokens[i].size() + 1;
       r->timestamps[i] = result.timestamps[i];
     }
+
+    r->tokensArr = const_cast<const char *const *>(tokensTemp);
   } else {
     r->count = 0;
     r->timestamps = nullptr;
     r->tokens = nullptr;
+    r->tokensArr = nullptr;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.cc]: Please also add code to free it to avoid memory leak.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: sherpa-onnx/c-api/c-api.cc (hunk 2 above)
Comment (ductranminh) [reply sherpa-onnx/c-api/c-api.cc]: @csukuangfj sorry, what do you want to free? This line is just to initialize null for pointer incase there is no token from `SherpaOnnxOnlineRecognizerResult`

Comment (csukuangfj) [reply sherpa-onnx/c-api/c-api.cc]: Please change
https://github.com/k2-fsa/sherpa-onnx/blob/48038a78e2bc0c56e72ce49973058c12f815b5bc/sherpa-onnx/c-api/c-api.cc#L179-L182
to
```cpp
void DestroyOnlineRecognizerResult(const SherpaOnnxOnlineRecognizerResult *r) {
  delete[] r->text;
  delete[] r->json;
  delete[] r->tokensArr;
  delete r;
}
```

---

By the way, please change the variable names:
 - `totalLength` -> `total_length`
 - `tokensTemp` -> `tokens_temp`
 - `tokensArr` -> `tokens_arr`
 to follow the code convention in sherpa-onnx.

File: sherpa-onnx/c-api/c-api.cc [hunk 3]
```diff
@@ -128,10 +128,51 @@ SherpaOnnxOnlineRecognizerResult *GetOnlineStreamResult(
...
+    for (int32_t i = 0; i < r->count; ++i) {
+      tokensTemp[i] = const_cast<char*>(r->tokens) + pos;
+      memcpy(reinterpret_cast<void *>(const_cast<char *>(r->tokens + pos)),
+             result.tokens[i].c_str(), result.tokens[i].size());
+      // +1 to move past the null character
+      pos += result.tokens[i].size() + 1;
+      r->timestamps[i] = result.timestamps[i];
+    }
+
+    r->tokensArr = const_cast<const char *const *>(tokensTemp);
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.cc]:
```suggestion
    r->tokensArr = tokensTemp;
```

Comment (ductranminh) [reply sherpa-onnx/c-api/c-api.cc]: I'm done

File: sherpa-onnx/c-api/c-api.cc [hunk 4]
```diff
@@ -128,15 +128,59 @@ SherpaOnnxOnlineRecognizerResult *GetOnlineStreamResult(
...
+  }
+
   return r;
 }
 
 void DestroyOnlineRecognizerResult(const SherpaOnnxOnlineRecognizerResult *r) {
   delete[] r->text;
+  delete[] r->json;
+  delete[] r->tokens;
+  delete[] r->tokens_arr;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.cc:183]: Please also delete
```cpp
delete[] r->timestamps;
```

File: sherpa-onnx/c-api/c-api.cc [hunk 5]
```diff
@@ -128,15 +128,59 @@ SherpaOnnxOnlineRecognizerResult *GetOnlineStreamResult(
...
+    r->count = count;
+    // Each word ends with nullptr
+    r->tokens = new char[total_length];
+    memset(reinterpret_cast<void *>(const_cast<char *>(r->tokens)), 0,
+           total_length);
+    r->timestamps = new float[r->count];
+    char **tokens_temp = new char*[r->count];
+    int pos = 0;
+    for (int32_t i = 0; i < r->count; ++i) {
+      tokens_temp[i] = const_cast<char*>(r->tokens) + pos;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.cc:160]:
```suggestion
      tokens_temp[i] = r->tokens + pos;
```

File: sherpa-onnx/c-api/c-api.cc [hunk 6]
```diff
@@ -128,15 +128,59 @@ SherpaOnnxOnlineRecognizerResult *GetOnlineStreamResult(
...
+    }
+
+    r->count = count;
+    // Each word ends with nullptr
+    r->tokens = new char[total_length];
+    memset(reinterpret_cast<void *>(const_cast<char *>(r->tokens)), 0,
+           total_length);
+    r->timestamps = new float[r->count];
+    char **tokens_temp = new char*[r->count];
+    int pos = 0;
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.cc]:
```suggestion
    int32_t pos = 0;
```

File: sherpa-onnx/c-api/c-api.h [hunk 7]
```diff
@@ -101,8 +101,36 @@ SHERPA_ONNX_API typedef struct SherpaOnnxOnlineRecognizerConfig {
...
+
+  // a pointer array contains the address of the first item in tokens
+  const char *const *tokens_arr;
+
+  // Pointer to continuous memory which holds timestamps which
+  // are seperated by \0
```
Comment (csukuangfj) [inline sherpa-onnx/c-api/c-api.h]: Please update the comment. It is not separated by \0 for timestamps.

File: sherpa-onnx/c-api/c-api.cc (hunk 5 above)
Comment (ductranminh) [reply sherpa-onnx/c-api/c-api.cc:160]: The reason for this casting `const_cast<char*>`  is build error:  
```
error: assigning to 'char *' from 'const char *' discards qualifiers
tokens_temp[i] = r->tokens + pos;
```

File: sherpa-onnx/c-api/c-api.h (hunk 7 above)
Comment (ductranminh) [reply sherpa-onnx/c-api/c-api.h]: fixed

**Comment by csukuangfj** [comment]
Thanks for your contribution!

**Pull request merged**